esp-hal/src/dma/buffers.rs

```diff
@@ -4,25 +4,252 @@ use core::{
+        // should be 16-byte, 32-byte or 64-byte aligned. For data frame whose
+        // length is not a multiple of 16 bytes, 32 bytes, or 64 bytes, EDMA adds
+        // padding bytes to the end.
+
+        // S3: Size and Address for IN transfers must be block aligned. For receive
+        // descriptors, if the data length received are not aligned with block size,
+        // GDMA will pad the data received with 0 until they are aligned to
+        // initiate burst transfer. You can read the length field in receive descriptors
+        // to obtain the length of valid data received
+        if matches!(direction, TransferDirection::In) {
+            self as usize
+        } else {
+            // S2: Size, length and buffer address pointer in transmit descriptors are not
+            // necessarily aligned with block size.
+
+            // S3: Size, length, and buffer address pointer in transmit descriptors do not
+            // need to be aligned.
+            1
+        }
+    }
+}
+
+impl InternalBurstTransfer {
     pub(super) fn is_burst_enabled(self) -> bool {
         !matches!(self, Self::Disabled)
     }
+
+    const fn min_dram_alignment(self, direction: TransferDirection) -> usize {
+        // IN transfers must be word aligned
+        if matches!(direction, TransferDirection::In) {
+            4
+        } else {
+            // OUT transfers have no alignment requirements, except for ESP32, which is
+            // described below.
+            if cfg!(esp32) {
+                // SPI DMA: Burst transmission is supported. The data size for
+                // a single transfer must be four bytes aligned.
+                // I2S DMA: Burst transfer is supported. However, unlike the
+                // SPI DMA channels, the data size for a single transfer is
+                // one word, or four bytes.
+                4
+            } else {
+                1
+            }
+        }
+    }
+}
+
+const fn max(a: usize, b: usize) -> usize {
+    if a > b {
+        a
+    } else {
+        b
+    }
+}
+
+impl BurstConfig {
+    delegate::delegate! {
+        #[cfg(psram_dma)]
+        to self.internal {
+            pub(super) const fn min_dram_alignment(self, direction: TransferDirection) -> usize;
+            pub(super) fn is_burst_enabled(self) -> bool;
+        }
+    }
+
+    /// Calculates an alignment that is compatible with the current burst
+    /// configuration.
+    ///
+    /// This is an over-estimation so that Descriptors can be safely used with
+    /// any DMA channel in any direction.
+    pub const fn min_compatible_alignment(self) -> usize {
+        let in_alignment = self.min_dram_alignment(TransferDirection::In);
+        let out_alignment = self.min_dram_alignment(TransferDirection::Out);
+        let alignment = max(in_alignment, out_alignment);
+
+        #[cfg(psram_dma)]
+        let alignment = max(alignment, self.external as usize);
```

Comment:
Technically max is the incorrect operator to use, you're assuming the alignments are multiples of each other. Someone will definitely miss this in future.

What you really want is "Least common multiple".

Reply:
At this point I think we're fine, future devices seem to relax alignment requirements anyway except where they can't. In every case, external alignment is a multiple of 8 or 16, and internal is 1 or 4.